### PR TITLE
Bring back support for `update_with_oracle_cut_size`

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This option was available in the pre-refactor parser, but was never implemented in the refactored parser. This option cuts transition sequences that are longer than `update_with_oracle_cut` size into separate sequences that have at most `update_with_oracle_cut` transitions. The oracle (gold standard) transition sequence is used to determine the cuts and the initial states for the additional sequences.

Applying this cut makes the batches more homogeneous in the transition sequence lengths, making forward passes (and as a consequence training) much faster.

Training time 1000 steps on `de_core_news_lg`:

- Before this change: 149s
- After this change: 68s
- Pre-refactor parser: 81s

Also fix up an unrelated issue where the parser rehearsal tests would fail. Since this PR affects rehearsal as well (in a minor way), I bundled up that fix here.

### Types of change

Performance improvement.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/ml/tb_framework.pyx [hunk 1]
```diff
@@ -148,32 +148,74 @@ def init(
...
+
+    def __init__(self, docs: List[Doc], moves: TransitionSystem,
+        actions: Optional[List[Ints1d]]=None, max_moves: int=0,
+        states: Optional[List[State]]=None):
+        """
+        actions (Optional[List[Ints1d]]): actions to apply for each Doc.
+        docs (List[Doc]): Docs to predict transition sequences for.
+        max_moves: (Optional[int]): the maximum number of moves to apply,
+            values less than 1 will apply moves to states until they are
+            final states.
```
Comment (shadeMe) [inline spacy/ml/tb_framework.pyx:182]: A couple of things:
- `max_moves` isn't an optional by virtue of accepting a `None`, so we should update the description.
- On a similar note, given that the semantics of a zero value and is the same as that of any negative value, we could make this more explicit by restricting the lower-bound to zero and raising an exception when we receive a negative value.

File: spacy/ml/tb_framework.pyx [hunk 2]
```diff
@@ -148,32 +148,74 @@ def init(
...
 
-    states = moves.init_batch(docs)
+    states = moves.init_batch(docs) if inputs.states is None else inputs.states
     tokvecs, backprop_tok2vec = tok2vec(docs, is_train)
     tokvecs = model.ops.xp.vstack((tokvecs, hidden_pad))
     feats, backprop_feats = _forward_precomputable_affine(model, tokvecs, is_train)
     seen_mask = _get_seen_mask(model)
 
-    # Fixme: support actions in forward_cpu
     if beam_width == 1 and not is_train and isinstance(model.ops, NumpyOps):
```
Comment (shadeMe) [inline spacy/ml/tb_framework.pyx:211]: Minor nitpick:
```suggestion
    if not is_train and beam_width == 1 and isinstance(model.ops, NumpyOps):
```

File: spacy/pipeline/transition_parser.pyx [hunk 3]
```diff
@@ -312,12 +332,15 @@ class Parser(TrainablePipe):
         del backprop_scores
         return losses
 
-    def get_loss(self, states_scores, examples):
-        states, scores = states_scores
+    def get_loss(self, states_scores, examples, max_moves):
```
Comment (shadeMe) [inline spacy/pipeline/transition_parser.pyx:335]: Missing type hints.

File: spacy/pipeline/transition_parser.pyx [hunk 4]
```diff
@@ -335,7 +358,7 @@ class Parser(TrainablePipe):
         d_scores -= (costs <= best_costs) * (exp_gscores / gZ)
         return d_scores
 
-    def _get_costs_from_histories(self, examples, histories):
+    def _get_costs_from_histories(self, examples, gold_states, init_states, histories, max_moves):
```
Comment (shadeMe) [inline spacy/pipeline/transition_parser.pyx:361]: Missing type hints.

File: spacy/pipeline/transition_parser.pyx [hunk 5]
```diff
@@ -502,6 +527,49 @@ class Parser(TrainablePipe):
                     raise ValueError(Errors.E149) from None
         return self
 
+    def _init_gold_batch(self, examples, max_length):
```
Comment (shadeMe) [inline spacy/pipeline/transition_parser.pyx:530]: Missing type hints.

File: spacy/pipeline/transition_parser.pyx [hunk 6]
```diff
@@ -502,6 +527,49 @@ class Parser(TrainablePipe):
...
+        """Make a square batch, of length equal to the shortest transition
+        sequence or a cap. A long
+        doc will get multiple states. Let's say we have a doc of length 2*N,
+        where N is the shortest doc. We'll make two states, one representing
+        long_doc[:N], and another representing long_doc[N:]."""
```
Comment (shadeMe) [inline spacy/pipeline/transition_parser.pyx:535]:
```suggestion
        sequence or a cap. A long doc will get multiple states. Let's say we have a doc 
        of length 2*N, where N is the shortest doc. We'll make two states, one representing
        long_doc[:N], and another representing long_doc[N:]."""
```
Whitespace.

File: spacy/pipeline/transition_parser.pyx (hunk 3 above)
Comment (danieldk) [reply spacy/pipeline/transition_parser.pyx:335]: I think we should do these in a separate PR, practically nothing in these modules have type hints. So I'd be more in favor of adding them all at once in a separate PR.

File: spacy/ml/tb_framework.pyx [hunk 7]
```diff
@@ -148,32 +148,74 @@ def init(
...
+        "actions",
+        "docs",
+        "max_moves",
+        "moves",
+        "states",
+    ]
+
+    def __init__(self, docs: List[Doc], moves: TransitionSystem,
+        actions: Optional[List[Ints1d]]=None, max_moves: int=0,
+        states: Optional[List[State]]=None):
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/ml/tb_framework.pyx:176]: What I wouldn't give for black for Python 😭 .

File: spacy/pipeline/transition_parser.pyx (hunk 3 above)
Comment (danieldk) [reply spacy/pipeline/transition_parser.pyx:335]: I’ll create a ticket for it tomorrow.

File: spacy/ml/tb_framework.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/ml/tb_framework.pyx]: Fixed.

File: spacy/ml/tb_framework.pyx (hunk 1 above)
Comment (danieldk) [reply spacy/ml/tb_framework.pyx]: Fixed `max_moves` description.

> On a similar note, given that the semantics of a zero value and is the same as that of any negative value, we could make this more explicit by restricting the lower-bound to zero and raising an exception when we receive a negative value.

The pre-refactor parser does not have this restriction, so adding this restriction now could break existing configurations and it doesn't give us much.

File: spacy/ml/tb_framework.pyx (hunk 2 above)
Comment (danieldk) [reply spacy/ml/tb_framework.pyx]: Fixed.

File: spacy/pipeline/transition_parser.pyx (hunk 3 above)
Comment (danieldk) [reply spacy/pipeline/transition_parser.pyx:334]: Created a ticket. I resolved the other type hints comments.

File: spacy/pipeline/transition_parser.pyx (hunk 6 above)
Comment (danieldk) [reply spacy/pipeline/transition_parser.pyx]: Fixed.

File: spacy/ml/tb_framework.pyx [hunk 8]
```diff
@@ -230,7 +272,7 @@ cdef list _parse_batch(CBlas cblas, TransitionSystem moves, StateC** states,
 
 
 def _forward_fallback(model: Model, moves: TransitionSystem, states: List[StateClass], tokvecs, backprop_tok2vec, feats, backprop_feats, seen_mask, is_train: bool,
-                  actions: Optional[List[Ints1d]]=None):
+        actions: Optional[List[Ints1d]]=None, max_moves=0):
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/ml/tb_framework.pyx]: Fixed for the newly-added argument.

**Review by shadeMe — APPROVED** [review]

File: spacy/ml/tb_framework.pyx [hunk 9]
```diff
@@ -281,6 +336,9 @@ def _forward_fallback(model: Model, moves: TransitionSystem, states: List[StateC
             all_ids.append(ids)
             all_statevecs.append(statevecs)
             all_which.append(which)
+        if max_moves >= 1 and n_moves >= max_moves:
```
Comment (svlandeg) [inline spacy/ml/tb_framework.pyx:339]: I'd personally prefer chaining this: 
```suggestion
        if n_moves >= max_moves >= 1:
```

**Review by svlandeg — COMMENTED** [review]
Looking good!

Can we add some unit tests here for the general behaviour, but also to test some edge cases like `max_moves` being 0 or 1, etc?

File: spacy/ml/tb_framework.pyx (hunk 9 above)
Comment (danieldk) [reply spacy/ml/tb_framework.pyx]: Fixed.

**Comment by danieldk** [comment]
> Can we add some unit tests here for the general behaviour, but also to test some edge cases like max_moves being 0 or 1, etc?

Extended the parser overfitting test to train with cut sizes 0, 1, 5, and 100. Found (and fixed) a bug with a cut size of 1.
